<div class="sheet">
    <div class="title_area">
      <h3>{{page_title}}</h3>
    </div>
</div>

<div class="out-sheet">
  <div id="examples_sheet" class="sheet" style="min-height: 700px;">
    <br />

<!-- -->
    <br />
    <br />

    <h4>For the calculation mstatspop, formats allowed:</h4>
    <br />
    <ul>
      <li><strong>Fasta: </strong>Using the fasta file it is only possible to make a whole analysis of the entire region but not a sliding window analysis.</li>
      <li><strong>MS: </strong>The calculation mstatspop allow to include ms format files. It is useful in case of making simulation analysis for specific regions.</li>
      <li><strong>T-fasta: </strong>The normal pipeline will use transposed fasta format (.tfa). This format contains the sequences for each individual in vertical instead of horizontal. It also includes a header with the names of each sequence and the number position of each nucleotide at the beginning of each row.</li>
    </ul>
    <br />
    <br />


            <div class="container">

              <div class="row">
                <div class="col-sm-11 col-lg-11">
                  <h4>FASTA EXAMPLES - WHOLE REGION ANALYSIS</h4>
                </div>
              </div>

<!--
              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">&lt;Description&gt;</span><br />
              ngasp mstatspop -f fasta -i ./100Kchr10.fa -o 0 -N 1 42   -T ./mstatspop_100chr10.fa.01.txt</div>
              </div>
-->
              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Same result but in a single line</span><br />
              ngasp mstatspop -f fasta -i ./100Kchr10.fa -o 1 -N 1 42   -T ./mstatspop_100chr10.fa.01b.txt</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two populations</span><br />
              ngasp mstatspop -f fasta -i ./100Kchr10.fa -o 0 -N 2 40 2 -T ./mstatspop_100chr10.fa.02.txt</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + outgroup</span><br />
              ngasp mstatspop -f fasta -i ./100Kchr10.fa -o 0 -N 2 40 2 -T ./mstatspop_100chr10.fa.03.txt -G 1</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + outg + missing</span><br />
              ngasp mstatspop -f fasta -i ./100Kchr10.fa -o 0 -N 2 40 2 -T ./mstatspop_100chr10.fa.04.txt -G 1 -u 1</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Three pops + outg + missing</span><br />
              ngasp mstatspop -f fasta -i ./100Kchr10.fa -o 0 -N 3 20 20 2 -T ./mstatspop_100chr10.fa.05.txt -G 1 -u 1</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + outg + missing + IUPAC diploid</span><br />
              ngasp mstatspop -f fasta -i ./100Kchr10.fa -o 0 -N 2 40 2 -T ./mstatspop_100chr10.fa.06.txt -G 1 -u 1 -p 2</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + outg + missing + GTF2 for nonsyn</span><br />
              ngasp mstatspop -f fasta -i ./100Kchr10.fa -o 0 -N 2 40 2 -T ./mstatspop_100chr10.fa.07.txt -G 1 -u 1 -g ./100Kchr10.gtf nonsynonymous Nuclear_Universal -c max</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Three pops + outg + missing + permutation test</span><br />
              ngasp mstatspop -f fasta -i ./100Kchr10.fa -o 0 -N 3 20 20 2 -T ./mstatspop_100chr10.fa.08.txt -G 1 -u 1 -t 1000 -s 1684</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Three pops + outg + missing + reordering the two pops in inverse order</span><br />
              ngasp mstatspop -f fasta -i ./100Kchr10.fa -o 0 -N 3 20 20 2 -T ./mstatspop_100chr10.fa.09.txt -G 1 -u 1 -O 42 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 40 41</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Three pops + outg + missing + GTF2 for nonsyn</dnegritaiv>
              ngasp mstatspop -f fasta -i ./100Kchr10.fa -o 0 -N 3 20 20 2 -T ./mstatspop_100chr10.fa.10.txt -G 1 -u 1  -g ./100Kchr10.gtf nonsynonymous Nuclear_Universal -c max</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Three pops + outg + missing + reordering the two pops in inverse order + GTF2 for nonsyn</span><br />
              ngasp mstatspop -f fasta -i ./100Kchr10.fa -o 0 -N 3 20 20 2 -T ./mstatspop_100chr10.fa.11.txt -G 1 -u 1 -O 42 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 40 41  -g ./100Kchr10.gtf nonsynonymous Nuclear_Universal -c max</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Optimal tests using non-missing data</span><br />
              ngasp mstatspop -f fasta -i ./MC1R_PigsOutg_aligned.fas -o 0 -p 1 -u 0 -t 1000 -s 123456 -G 1 -N 3 48 46 1 -a ./MC1R_H1frq.txt -g ./MC1R.gff nonsynonymous Nuclear_Universal -T ./MC1R_PigsOutg_NSyn_Opttest.txt</div>
              </div>


              <div class="row">
                <div class="col-sm-11 col-lg-11">
                  <h4>MS EXAMLPES - SIMULATION ANALYSIS OF A SINGLE REGION</h4>
                </div>
              </div>


              <div class="row">
              <div class="col-sm-11 col-lg-11">
<!--          <span class="little_title">&lt;Description&gt;</span><br />
-->
              ngasp mstatspop -f ms -i ./100Kchr10_tfa2ms_02.ms.txt -o 0 -N 1 42 -T ./mstatspop_100chr10.ms.00.txt -l 100000</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11">
<!--          <span class="little_title">&lt;Description&gt;</span><br />
-->
              ngasp mstatspop -f ms -i ./100Kchr10_tfa2ms_02.ms.txt -o 0 -N 1 42 -T ./mstatspop_100chr10.ms.01.txt -l 100000 -m ./100Kchr10_npop1_nsam42_ExcludeMissingVariantsmhits_NOoutg_MASK.txt</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Same result but in a single line</span><br />
              ngasp mstatspop -f ms -i ./100Kchr10_tfa2ms_02.ms.txt -o 1 -N 1 42 -T ./mstatspop_100chr10.ms.01b.txt -l 100000 -m ./100Kchr10_npop1_nsam42_ExcludeMissingVariantsmhits_NOoutg_MASK.txt</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two populations</span><br />
              ngasp mstatspop -f ms -i ./100Kchr10_tfa2ms_03.ms.txt -o 0 -N 2 40 2 -T ./mstatspop_100chr10.ms.02.txt -l 100000 -m ./100Kchr10_npop2_nsam42_ExcludeMissingVariantsmhits_NOoutg_MASK.txt</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + outgroup</span><br />
              ngasp mstatspop -f ms -i ./100Kchr10_tfa2ms_03.ms.txt -o 0 -N 2 40 2 -T ./mstatspop_100chr10.ms.03.txt -l 100000 -G 1 -m ./100Kchr10_npop1_nsam42_ExcludeMissingVariantsmhits_NOoutg_MASK.txt</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">One pop + several ms iterations</span><br />
              ngasp mstatspop -f ms -i ./100Kchr10_tfa2ms_07.ms.txt -o 1 -N 1 42 -T ./mstatspop_100chr10.ms.04.txt -l 10000 -r 5</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + several ms iterations</span><br />
              ngasp mstatspop -f ms -i ./100Kchr10_tfa2ms_06.ms.txt -o 1 -N 2 40 2 -T ./mstatspop_100chr10.ms.05.txt -l 10000 -r 5</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + several ms iterations + forcing outgroup the value '0'</span><br />
              ngasp mstatspop -f ms -i ./100Kchr10_tfa2ms_06.ms.txt -o 1 -N 2 40 2 -T ./mstatspop_100chr10.ms.06.txt -l 10000 -r 5 -F 1</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + several ms iterations  + forcing outgroup the value '0' + revert mutation </span><br />
              ngasp mstatspop -f ms -i ./100Kchr10_tfa2ms_06.ms.txt -o 1 -N 2 40 2 -T ./mstatspop_100chr10.ms.07.txt -l 10000 -r 5 -F 1 -q 0.01</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + outg + several ms iterations + s/v ratio</span><br />
              ngasp mstatspop -f ms -i ./100Kchr10_tfa2ms_06.ms.txt -o 1 -N 2 40 2 -T ./mstatspop_100chr10.ms.08.txt -l 10000 -r 5 -G 1 -v 2.0</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + outg + missing</span><br />
              ../bin/ms 42 1 -t 1000 -I 2 40 2 -ej 2.0 2 1 -p 9 > ./100Kchr10_simulation.ms.txt</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11">
<!--          <span class="little_title">&lt;Description&gt;</span><br />
-->
              ngasp mstatspop -f ms -i ./100Kchr10_simulation.ms.txt -o 0 -N 2 40 2 -T ./mstatspop_100chr10.ms.09.txt -l 100000 -G 1 -u 1 -m ./100Kchr10_npop2_nsam42_ExcludeMissingVariantsmhits_NOoutg_MASK.txt</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + outg + missing + weight for nonsyn</span><br />
              ngasp mstatspop -f ms -i ./100Kchr10_simulation.ms.txt -o 0 -N 2 40 2 -T ./mstatspop_100chr10.ms.10.txt -l 100000 -G 1 -u 1 -m ./100Kchr10_npop2_nsam42_nonsynonymous_max_IncludeMissingVariantsmhits_outg_MASK.txt</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + outg + missing + weight for nonsyn + permutation test</span><br />
              ngasp mstatspop -f ms -i ./100Kchr10_simulation.ms.txt -o 0 -N 2 40 2 -T ./mstatspop_100chr10.ms.11.txt -l 100000 -G 1 -u 1 -m ./100Kchr10_npop2_nsam42_nonsynonymous_max_IncludeMissingVariantsmhits_outg_MASK.txt -t 1000 -s 1684</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + outg + missing + weight for nonsyn + permutation test + s/v ratio</span><br />
              ngasp mstatspop -f ms -i ./100Kchr10_simulation.ms.txt -o 0 -N 2 40 2 -T ./mstatspop_100chr10.ms.12.txt -l 100000 -G 1 -u 1 -m ./100Kchr10_npop2_nsam42_nonsynonymous_max_IncludeMissingVariantsmhits_outg_MASK.txt -t 1000 -s 1684 -v 2.0</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + outg + missing + weight for nonsyn + permutation test + s/v ratio + 1000 replicates</span><br />
                  ../bin/ms 42 1000 -t 1000 -I 2 40 2 -ej 2.0 2 1 -p 9 &gt; ./100Kchr10_simulation1000.ms.txt<br />
                  ngasp mstatspop -f ms -i ./100Kchr10_simulation1000.ms.txt -o 1 -N 2 40 2 -T ./mstatspop_100chr10.ms.13.txt -l 100000 -G 1 -u 1 -m ./100Kchr10_npop2_nsam42_nonsynonymous_max_IncludeMissingVariantsmhits_outg_MASK.txt -t 1000 -s 1684 -v 2.0 -r 1000</div>
              </div>

              <div class="row">
                <div class="col-sm-11 col-lg-11">
                  <h4>T-FASTA EXAMPLES - SLIDING WINDOW ANALYSIS OF EMPIRICAL DATA</h4>
                </div>
              </div>


              <div class="row">
              <div class="col-sm-11 col-lg-11">
<!--          <span class="little_title">&lt;Description&gt;</span><br />
-->
              ngasp mstatspop -f tfa -i ./100Kchr10.tfa -o 0 -N 1 42   -T ./mstatspop_100chr10.tfa.01.txt -w 100000</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Same result but in a single line</span><br />
              ngasp mstatspop -f tfa -i ./100Kchr10.tfa -o 1 -N 1 42   -T ./mstatspop_100chr10.tfa.01b.txt -w 100000</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two populations</span><br />
              ngasp mstatspop -f tfa -i ./100Kchr10.tfa -o 0 -N 2 40 2 -T ./mstatspop_100chr10.tfa.02.txt -w 100000</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + outgroup</span><br />
              ngasp mstatspop -f tfa -i ./100Kchr10.tfa -o 0 -N 2 40 2 -T ./mstatspop_100chr10.tfa.03.txt -G 1 -w 100000</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + outg + missing</span><br />
              ngasp mstatspop -f tfa -i ./100Kchr10.tfa -o 0 -N 2 40 2 -T ./mstatspop_100chr10.tfa.04.txt -G 1 -u 1 -w 100000</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + outg + missing</span><br />
              ngasp mstatspop -f tfa -i ./10bpchr10.tfa -o 10 -N 2 40 2 -T ./mstatspop_10bchr10.tfa.04.txt -G 1 -u 1 -w 10</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + outg + missing</span><br />
              ngasp mstatspop -f tfa -i ./10bpchr10.tfa -o 6 -N 2 40 2 -T ./mstatspop_10bchr10c.tfa.04.txt -G 1 -u 1 -w 10</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Three pops + outg + missing</span><br />
              ngasp mstatspop -f tfa -i ./100Kchr10.tfa -o 0 -N 3 20 20 2 -T ./mstatspop_100chr10.tfa.05.txt -G 1 -u 1 -w 100000</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + outg + missing + window 10000</span><br />
              ngasp mstatspop -f tfa -i ./100Kchr10.tfa -o 1 -N 2 40 2 -T ./mstatspop_100chr10.tfa.06.txt -G 1 -u 1 -w 10000</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + outg + missing + window 10000 + slide 20000</span><br />
              ngasp mstatspop -f tfa -i ./100Kchr10.tfa -o 1 -N 2 40 2 -T ./mstatspop_100chr10.tfa.07.txt -G 1 -u 1 -w 10000 -z 20000</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + outg + missing + coordinates. Equal than the previous example</span><br />
              ngasp mstatspop -f tfa -i ./100Kchr10.tfa -o 0 -N 2 40 2 -T ./mstatspop_100chr10.tfa.08.txt -G 1 -u 1 -W ./coord_100Kb.txt -w 100000</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + outg + missing + window 10000 + weight for nonsyn</span><br />
              ngasp mstatspop -f tfa -i ./100Kchr10.tfa -o 0 -N 2 40 2 -T ./mstatspop_100chr10.tfa.09.txt -G 1 -u 1 -w 10000 -E ./100Kchr10_fa2tfa_08.tfa_npops2_nsam42_nonsynonymous_max_IncludeMissingVariantsmhits_outg_ploidy1_WEIGHTS.txt</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + outg + missing + window 1000 + weight for nonsyn + Effective_positions</span><br />
              ngasp mstatspop -f tfa -i ./100Kchr10.tfa -o 0 -N 2 40 2 -T ./mstatspop_100chr10.tfa.10.txt -G 1 -u 1 -w 1000 -E ./100Kchr10_fa2tfa_08.tfa_npops2_nsam42_nonsynonymous_max_IncludeMissingVariantsmhits_outg_ploidy1_WEIGHTS.txt -Y 0</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + outg + missing + weight for nonsyn + Effective_positions</span><br />
              ngasp mstatspop -f tfa -i ./100Kchr10.tfa -o 0 -N 2 40 2 -T ./mstatspop_100chr10.tfa.11.txt -G 1 -u 1 -w 100000 -E ./100Kchr10_fa2tfa_08.tfa_npops2_nsam42_nonsynonymous_max_IncludeMissingVariantsmhits_outg_ploidy1_WEIGHTS.txt -Y 0</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Three pops + outg + missing + reordering the two pops in inverse order</span><br />
              ngasp mstatspop -f tfa -i ./100Kchr10.tfa -o 0 -N 3 20 20 2 -T ./mstatspop_100chr10.tfa.12.txt -G 1 -u 1 -O 42 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 40 41 -w 100000</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Three pops + outg + missing + weight for nonsyn</span><br />
              ngasp mstatspop -f tfa -i ./100Kchr10.tfa -o 0 -N 3 20 20 2 -T ./mstatspop_100chr10.tfa.13.txt -G 1 -u 1  -w 100000 -E ./100Kchr10_fa2ms_15.ms.txt_npops3_nsam84_nonsynonymous_max_ExcludeMissingmhits_outg_ploidy2_WEIGHTS.txt</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Three pops + outg + missing + reordering the two pops in inverse order + weight for nonsyn</span><br />
              ngasp mstatspop -f tfa -i ./100Kchr10.tfa -o 0 -N 3 20 20 2 -T ./mstatspop_100chr10.tfa.14.txt -G 1 -u 1  -w 100000 -O 42 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 40 41  -E ./100Kchr10_fa2ms_15.ms.txt_npops3_nsam84_nonsynonymous_max_ExcludeMissingmhits_outg_ploidy2_WEIGHTS.txt</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Three pops + outg + missing + reordering the two pops in inverse order + weight for nonsyn + Effective positions</span><br />
              ngasp mstatspop -f tfa -i ./100Kchr10.tfa -o 0 -N 3 20 20 2 -T ./mstatspop_100chr10.tfa.14b.txt -G 1 -u 1  -w 100000 -O 42 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 40 41  -E ./100Kchr10_fa2ms_15.ms.txt_npops3_nsam84_nonsynonymous_max_ExcludeMissingmhits_outg_ploidy2_WEIGHTS.txt -Y 0</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + outg + missing + window 10000 + slide 20000 + permutations: like example 07</span><br />
              ngasp mstatspop -f tfa -i ./100Kchr10.tfa -o 1 -N 2 40 2 -T ./mstatspop_100chr10.tfa.15.txt -G 1 -u 1 -w 10000 -z 20000 -t 1000 -s 1684</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Two pops + outg + missing + window 100 + slide 100 + permutations:</span><br />
              ngasp mstatspop -f tfa -i ./100Kchr10.tfa -o 1 -N 2 40 2 -T ./mstatspop_100chr10.tfa.16.txt -G 1 -u 1 -w 100 -z 100 -t 1000 -s 1684</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Three pops + outg + missing + window 100 + slide 100 + permutations:</span><br />
              ngasp mstatspop -f tfa -i ./100Kchr10.tfa -o 1 -N 3 20 20 2 -T ./mstatspop_100chr10.tfa.17.txt -G 1 -u 1 -w 100 -z 100 -t 1000 -s 1684</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Three pops LESS SAMPLES + outg + missing + window 100 + slide 100 :</span><br />
              ngasp mstatspop -f tfa -i ./100Kchr10.tfa -o 1 -N 3 10 10 2 -T ./mstatspop_100chr10.tfa.18.txt -G 1 -u 1 -w 100 -z 100</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Three pops LESS SAMPLES + reordering + outg + missing + window 100 + slide 100 :</span><br />
              ngasp mstatspop -f tfa -i ./100Kchr10.tfa -o 1 -N 3 10 10 2 -T ./mstatspop_100chr10.tfa.19.txt -G 1 -u 1 -w 100 -z 100 -O 22 20 21 22 23 24 25 26 27 28 29 0 1 2 3 4 5 6 7 8 9 40 41</div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><h4>collect_data_columns.pl EXAMPLES</h4></div>
              </div>

              <div class="row">
              <div class="col-sm-11 col-lg-11"><span class="little_title">Collect some columns from the mstatspop output file</span><br />
                 perl collect_data_columns.pl -in ./mstatspop_100chr10.tfa.15.txt -fc ./required_columns1.txt > ./mstatspop_100chr10.tfa.15_COLUMNS.txt</div>
              </div>

			      </div>


          <br />
          <br />
          <br />
          <br />

          <br />
          <br />
          <br />
          <br />

<!-- -->
  </div>
</div>
