<div class="sheet">
  <div class="title_area">
      <h3>{{page_title}}</h3>
  </div>
</div>

<div class="out-sheet">
  <div class="sheet">
    <br />

    <p><a href="#" ng-click="goTo('downloads')">Download the ngaSP application</a>.</p>
    <br />
    <p>Depending on your initial datasets you can use a different pipeline:</p>
    <br />
    <ul>
    <li>
    <p><strong>BAM diploid individuals</strong></p>
    <br />
    <p>If you have individual diploid BAM files, the <a href="#" ng-click="goTo('pipeline')">pipeline 1</a> is suitable for your analysis. You will need the reference sequence, and optional outgroup sequence(s) and/or annotation file (GTF v2).</p>
    <br />
    <p>Right now, the pipeline works independently for each chromosome, so you will have to run each chromosome separately (one run of this pipeline per chromosome).</p>
    <br />
    <p>You can optionally analyze a number of different regions, chose the window size, chose the regions of the genome to study, choose the functional regions (synonymous, non-synonymous, silent, coding, etc…) you want to analyze. If you want to analyze many different options (e.g., synonymous and non-synonymous, different size of the sliding windows…), it is posible to keep intermediate results (that is, tfasta files) to run again a subset of the pipeline.</p>
    <br />
    </li>
    <li>
    <p><strong>Other datasets:</strong></p>
    <br />
    <p><strong>gVCF files</strong></p>
    <br />
    <p><strong>BAM files from pooled data</strong></p>
    </li>
    </ul>
    <br />
    <p>We are working in opening a number of different pipelines in order to work with gVCF and also with pooled data. Right now it is possible to analyze this kind of datasets using available related software.</p>


  </div>
</div>
